File: Practice_2 (Java)/src/ua/KarpovIvan/atd/DoublyLinkedList.java
package ua.KarpovIvan.atd;

import java.util.Iterator;

/*
 * Клас, що реалізує двоспрямований список.
 */
public class DoublyLinkedList<T> implements Iterable<T> {
    private DoublyNode<T> head; /* Початок списку. */
    private DoublyNode<T> tail; /* Кінець списку. */
    private int count = 0;      /* Кількість елементів в списку. */

    /*
     * Додавання елемента в початок списку.
     */
    public void addFirst(T data) {
        DoublyNode<T> node = new DoublyNode<>(data);
        DoublyNode<T> temp = head;

        node.setNext(temp);
        head = node;

        if (listEmpty()) {
            tail = head;
        }
        else {
            temp.setPrevious(node);
        }

        count++;
    }

    /*
     * Видалення елемента з кінця списку.
     */
    public void removeLast() throws Exception {
        DoublyNode<T> last = tail;

        if (listEmpty()) {
            throw new Exception("Список порожнiй.");
        }
        else if (count == 1) {
            head = null;
            tail = null;
            count = 0;
        }
        else {
            tail = last.getPrevious();
            tail.setNext(null);
            --count;
        }
    }

    /*
     * В завданні до практики сказано створити індексатор,
     * але засобами Java це неможливо зробити, тому я написав
     * наступні два методи.
     *
     * Встановлення елемента за індексом.
     */
    public void setElementByIndex(T value, int index) throws Exception {
        /* Перевірка, чи не виходе заданий індекс за допустимі межі. */
        if (index < 0 || index > count) {
            throw new Exception("Iндекс поза межами дiапазону!");
        }
        /* Якщо список порожній, елемент додається на початок. */
        else if (listEmpty()) {
            addFirst(value);
        }
        /* Додавання елемента по індексу. */
        else {
            DoublyNode<T> current = head;
            for (int i = 0; i < index; i++) {
                current = current.getNext();
            }
            current.setData(value);
        }
    }

    /*
     * Отримання елемента за індексом.
     */
    public T getElementByIndex(int index) throws Exception {
        /* Перевірка, чи не виходе заданий індекс за допустимі межі. */
        if (index < 0 || index > count) {
            throw new Exception("Iндекс поза межами дiапазону!");
        }

        /* Отримання значення елементу за індексом. */
        DoublyNode<T> current = head;
        for (int i = 0; i < index; i++) {
            current = current.getNext();
        }
        return current.getData();
    }

    /*
     * В завданні практики сказано створити властивість читання довжини списку,
     * але засобами Java це неможливо створити, тому це звичайний геттер.
     *
     * Метод отримання довини списку.
     */
    public int getCount() {
        return count;
    }

    /*
     * Перевірка на порожність списку.
     */
    public boolean listEmpty() { return count == 0; }

    /*
     * Метод, який повертає ітератор і довзоля проходитись
     * по списку в циклі foreach.
     */
    @Override
    public Iterator<T> iterator() {
        return new DoublyLinkedListIterator<>(head);
    }

    /*
     * Клас, який реалізує ітератор, за допомогою якого
     * відбувається ітерування по елементам списку.
     */
    public static class DoublyLinkedListIterator<E> implements Iterator<E> {
        DoublyNode<E> current;

        public DoublyLinkedListIterator(DoublyNode<E> head) {
            current = head;
        }

        @Override
        public boolean hasNext() {
            return current != null;
        }

        @Override
        public E next() {
            E value = current.getData();
            current = current.getNext();
            return value;
        }
    }

    /*
     * Метод, який повертає ітератор і довзоля проходитись
     * по списку в зворотньому порядку.
     */
    public Iterator<T> backIterator() {
        return new BackDoublyLinkedListIterator<>(tail);
    }

    /*
     * Клас, який реалізує ітератор, за допомогою якого
     * відбувається ітерування по елементам списку в зворотньому порядку.
     */
    public static class BackDoublyLinkedListIterator<E> implements Iterator<E> {
        DoublyNode<E> current;

        public BackDoublyLinkedListIterator(DoublyNode<E> tail) {
            current = tail;
        }

        @Override
        public boolean hasNext() {
            return current != null;
        }

        @Override
        public E next() {
            E value = current.getData();
            current = current.getPrevious();
            return value;
        }
    }

    /*
     * Допоміжний метод зміни місцями двох значень елементів списку.
     */
    public void swap(DoublyLinkedList<T> list, int i, int j) throws Exception {
        try {
            T temp = list.getElementByIndex(i);
            list.setElementByIndex(list.getElementByIndex(j), i);
            list.setElementByIndex(temp, j);
        }
        catch (Exception ex) {
            throw new Exception(ex);
        }
    }

    /*
     * Метод сортування списку за спаданням методом шейкер-сортування.
     */
    public void sort() throws Exception {
        /* Перевірка чи не порожній список. */
        if (listEmpty()) {
            throw new Exception("Список порожнiй!");
        }

        /* Якщо елементи в списку не типу Student, сортування не буде відбуватись. */
        if (getElementByIndex(0) instanceof Student) {
            int left = 0;
            int right = count - 1;

            /* Сортування. */
            while (left < right) {
                for (int i = left; i < right; i++) {
                    if (((Student)getElementByIndex(i)).getHeight() < ((Student)getElementByIndex(i + 1)).getHeight()) {
                        swap(this, i, i + 1);
                    }
                }
                right--;

                for (int i = right; i > left; i--) {
                    if (((Student)getElementByIndex(i - 1)).getHeight() < ((Student)getElementByIndex(i)).getHeight()) {
                        swap(this, i - 1, i);
                    }
                }
                left++;
            }
        }
    }

    /*
     * Метод пошуку студентів відмінників з ідеальною вагою.
     */
    public DoublyLinkedList<T> searchOfElement() throws Exception {
        /* Перевірка чи не порожній список. */
        if (listEmpty()) {
            throw new Exception("Список порожнiй!");
        }

        /*
         * Якщо елементи в списку не типу Student, пошук не буде відбуватись.
         * В такому випадку повертається порожній список. Якщо збігів не має,
         * також повертається пустий список.
         */
        if (getElementByIndex(0) instanceof Student) {
            DoublyNode<T> current = head;
            var searchResult = new DoublyLinkedList<T>();

            while (current != null) {
                boolean idealWeight = ((Student)current.getData()).getHeight() - 110 == ((Student)current.getData()).getWeight();
                if (idealWeight) {
                    searchResult.addFirst(current.getData());
                }
                current = current.getNext();
            }

            /* Повернення значення студента/ів, якщо був збіг. */
            if (!searchResult.listEmpty()) {
                return searchResult;
            }
        }
        return new DoublyLinkedList<>();
    }
}

File: Practice_1 (C#)/ListLib/DoublyLinkedList.cs
using System.Collections.Generic;
using System.Collections;
using System;

namespace ListLib
{
    // Клас, що реалізує двоспрямований список.
    public class DoublyLinkedList<T> : IEnumerable<T>
    {
        private DoublyNode<T> _head; // Початок списку.
        private DoublyNode<T> _tail; // Кінець списку.
        private int _count = 0;      // Кількість елементів в списку.

        // Додавання елемента в початок списку.
        public void AddFirst(T data)
        {
            DoublyNode<T> node = new DoublyNode<T>(data);
            DoublyNode<T> temp = _head;

            node.Next = temp;
            _head = node;

            if (ListEmpty)
            {
                _tail = _head;
            }
            else
            {
                temp.Previous = node;
            }

            _count++;
        }

        // Видалення елемента з кінця списку.
        public void RemoveLast()
        {
            DoublyNode<T> last = _tail;

            if (ListEmpty)
            {
                throw new Exception("Список порожній.");
            }
            else if (_count == 1)
            {
                _head = null;
                _tail = null;
                _count = 0;
            }
            else
            {
                _tail = last.Previous;
                _tail.Next = null;
                --_count;
            }
        }

        // Індексатор читання\зміни значення, заданого порядковим номером вузла.
        public T this[int index]
        {
            get
            {
                // Перевірка, чи не виходе заданий індекс за допустимі межі.
                if (index < 0 || index > _count)
                {
                    throw new IndexOutOfRangeException("Індекс поза межами діапазону!");
                }

                // Отримання значення елементу за індексом.
                DoublyNode<T> current = _head;
                for (int i = 0; i < index; i++)
                {
                    current = current.Next;
                }
                return current.Data;
            }
            set
            {
                // Перевірка, чи не виходе заданий індекс за допустимі межі.
                if (index < 0 || index > _count)
                {
                    throw new IndexOutOfRangeException("Індекс поза межами діапазону!");
                }
                // Якщо список порожній, елемент додається на початок.
                else if (ListEmpty)
                {
                    AddFirst(value);
                }
                // Додавання елемента по індексу.
                else
                {
                    DoublyNode<T> current = _head;
                    for (int i = 0; i < index; i++)
                    {
                        current = current.Next;
                    }
                    current.Data = value;
                }
            }
        }

        // Властивість довжини списку.
        public int Count { get => _count; }

        // Перевірка на порожність списку.
        public bool ListEmpty => _count == 0;

        // Метод ітерації списку.
        IEnumerator IEnumerable.GetEnumerator()
        {
            return ((IEnumerable)this).GetEnumerator();
        }

        // Метод ітерації списку для наступного елемента.
        IEnumerator<T> IEnumerable<T>.GetEnumerator()
        {
            DoublyNode<T> current = _head;
            while (current != null)
            {
                yield return current.Data;
                current = current.Next;
            }
        }

        // Метод ітерації списку в зворотньому порядку.
        public IEnumerable<T> BackEnumerator()
        {
            DoublyNode<T> current = _tail;
            while (current != null)
            {
                yield return current.Data;
                current = current.Previous;
            }
        }

        // Допоміжний метод зміни місцями двох значень елементів списку.
        public void Swap(DoublyLinkedList<T> list, int i, int j)
        {
            T temp = list[i];
            list[i] = list[j];
            list[j] = temp;
        }

        // Метод сортування списку за спаданням методом шейкер-сортування.
        public void Sort()
        {
            // Перевірка чи не порожній список.
            if (ListEmpty)
            {
                throw new Exception("Список порожній!");
            }

            // Якщо елементи в списку не типу Student, сортування не буде відбуватись.
            if (this[0] is Student)
            {
                int left = 0;
                int right = _count - 1;

                // Сортування.
                while (left < right)
                {
                    for (int i = left; i < right; i++)
                    {
                        if ((this[i] as Student).Gpa < (this[i + 1] as Student).Gpa)
                        {
                            Swap(this, i, i + 1);
                        }
                    }
                    right--;

                    for (int i = right; i > left; i--)
                    {
                        if ((this[i - 1] as Student).Gpa < (this[i] as Student).Gpa)
                        {
                            Swap(this, i - 1, i);
                        }
                    }
                    left++;
                }
            }
        }

        // Метод пошуку студентів відмінників що займаються спортом.
        public DoublyLinkedList<T> SearchOfElement()
        {
            // Перевірка чи не порожній список.
            if (ListEmpty)
            {
                throw new Exception("Список порожній!");
            }

            // Якщо елементи в списку не типу Student, нічого не буде шукатись,
            // а також повернеться стандатне значення для T.
            if (this[0] is Student)
            {
                DoublyNode<T> current = _head;
                var searchResult = new DoublyLinkedList<T>();

                while (current != null)
                {
                    if ((current.Data as Student).Gpa >= 90 && (current.Data as Student).Sport == true)
                    {
                        searchResult.AddFirst(current.Data);
                    }
                    current = current.Next;
                }
                
                // Повертає значення студентів, якщо був збіг, інакше: "Елемент не знайдено!".
                if (searchResult != null)
                {
                    return searchResult;
                }
                else
                {
                    throw new Exception("Елемент не знайдено!");
                }
            }
            return default;
        }
    }
}